js/bundle.go

```diff
@@ -58,18 +60,109 @@ type Bundle struct {
 	registry          *metrics.Registry
 
 	exports map[string]goja.Callable
+
+	cache    map[string]moduleCacheElement
+	reverse  map[goja.ModuleRecord]*url.URL
+	compiler *compiler.Compiler
 }
 
 // A BundleInstance is a self-contained instance of a Bundle.
 type BundleInstance struct {
-	Runtime *goja.Runtime
+	Runtime        *goja.Runtime
+	ModuleInstance goja.ModuleInstance
 
 	// TODO: maybe just have a reference to the Bundle? or save and pass rtOpts?
 	env map[string]string
 
 	exports      map[string]goja.Callable
 	moduleVUImpl *moduleVUImpl
-	pgm          programWithSource
+}
+
+func (bi *BundleInstance) getExported(name string) goja.Value {
+	return bi.ModuleInstance.GetBindingValue(unistring.NewFromString(name))
+}
+
+type moduleCacheElement struct {
+	err error
+	m   goja.ModuleRecord
+}
+
+// TODO use the first argument
+func (b *Bundle) resolveModule(ref interface{}, specifier string) (goja.ModuleRecord, error) {
```

Comment:
It's a very long method and a bit difficult to follow/review, can we split it a bit?